docs/source/tutorial_torch_generator_agent.md

```diff
@@ -0,0 +1,156 @@
+
+In `build_model()`, we instantiate our example model (defined below) by passing in the agent's dict (set by `TorchAgent`) and the hidden size. We add lines to optionally copy pre-existing token embeddings into the model's embedding table.
+
+Altogether, our example agent is defined as follows:
+
+```
+import parlai.core.torch_generator_agent as tga
+
+
+class Seq2seqAgent(tga.TorchGeneratorAgent):
+
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        super(Seq2seqAgent, cls).add_cmdline_args(argparser)
+        group = argparser.add_argument_group('Example TGA Agent')
+        group.add_argument(
+            '-hid', '--hidden-size', type=int, default=1024, help='Hidden size.'
+        )
+
+    def build_model(self):
+        model = ExampleModel(self.dict, self.opt['hidden_size'])
+        if self.opt['embedding_type'] != 'random':
+            self._copy_embeddings(
+                model.embeddings.weight, self.opt['embedding_type']
+            )
+        return model
+```
+
+### Extending `TorchGeneratorModel`
+
+We now subclass `TorchGeneratorModel` to create `ExampleModel`. We initialize this by first calling `super().__init__()` and passing in dictionary tokens for padding, start, end, and UNKs; we then create an embedding lookup table with `nn.Embedding` and instantiate the encoder and decoder, described in the following sections.
+
+```
+import torch.nn as nn
+import torch.nn.functional as F
+
+class ExampleModel(tga.TorchGeneratorModel):
+
+    def __init__(self, dictionary, esz=256, hidden_size=1024):
+        super().__init__(
+            padding_idx=dictionary[dictionary.null_token],
+            start_idx=dictionary[dictionary.start_token],
+            end_idx=dictionary[dictionary.end_token],
+            unknown_idx=dictionary[dictionary.unk_token],
+        )
+        self.embeddings = nn.Embedding(len(dictionary), esz)
+        self.encoder = Encoder(self.embeddings, hidden_size)
+        self.decoder = Decoder(self.embeddings, hidden_size)
+```
+
+We next define a function to project the output of the decoder back into the token space:
+
+```
+    def output(self, decoder_output):
+        return F.linear(decoder_output, self.embeddings.weight)
+```
+
+Lastly, we define two functions to reindex the latent states of the encoder and decoder along the sample dimension. We reindex the encoder at the very beginning of beam search and when ranking candidates during eval, and we reindex the decoder after each step of beam search. Since our encoder and decoder both are based on LSTMs, these encoder/decoder states are the hidden and cell states:
+```
+    def reorder_encoder_states(self, encoder_states, indices):
```

Comment:
can you add a comment explaining what the indices passed in here are?

Reply:
Yeah, that was vague - I added this